### ISSUE TITLE
Can't edit an existing link using only the keyboard

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#8266]
In Gutenberg 2.9.0 it was possible to;
- select an existing link in the content

<img width="830" alt="screen shot 2018-07-28 at 18 08 03" src="https://user-images.githubusercontent.com/1682452/43358431-f01024cc-9291-11e8-8bca-96559c2ecb85.png">

- press Tab and focus was moved on the link dialog: this way it was possible to edit a link using only the keyboard

<img width="842" alt="screen shot 2018-07-28 at 18 10 12" src="https://user-images.githubusercontent.com/1682452/43358442-19feeb7e-9292-11e8-8eea-3de998ec289a.png">

This worked because the link toolbar was placed "inline" right after the block editable area. Pressing tab followed the normal tab order, moving focus to the following focusable element in the native tab sequence.

After the changes in #6911 this is no longer true. The link toolbar is now placed within a "popover" component and thus it's at the end of the markup, way far from the block content. The link toolbar is not the next following focusable element any longer.

Right now I can't find a way to edit an existing link using only the keyboard:
- pressing Tab takes me elsewhere
- pressing Cmd+K makes the link toolbar appear but it's in "insert mode" and doesn't allow me to modify the existing link

I'd like to remind everyone that so important changes should be marked with the Accessibility label and tested extensively _also with keyboard only_ to make sure there are no regression.

I'd suggest to consider to replicate the Classic editor behavior. Pressing Cmd+K on an existing link should make the link toolbar appear (and get focused) in edit mode.

I've checked only the paragraph block, this should be checked for all the blocks that use the link toolbar.

Aside:
there's one more bit missing comparing the behavior with the Classic editor:
- in Gutenberg there's the need to select the whole linked word(s)
- in the Classic editor, placing the cursor within the linked word(s) is enough to edit the whole link

**Comment by abrightclearweb** [comment]
I noticed this today too.

Here is a gif of me creating a link and then trying to navigate back to it and edit it.
![editing a link keyboard only](https://user-images.githubusercontent.com/7430856/45307004-dce90700-b515-11e8-8c94-8869ba750e35.gif)

When I get the cursor within a link, I can press Shift+Tab to show the link popover, but then can't navigate into it. Pressing up, down, left or right just moves me right out.

The only way seems to be to use the Link button again to unlink the link, then recreate it. But that removes your link URL so not good.

Also, you can navigate to the "Open link in new tab" when you create a link but you can't toggle the option on until you have added a link already. So it is not much use if you can't go back and edit.

**Comment by gziolo** [comment]
There is no corresponding PR for the issue, moving to 4.2.

**Comment by afercia** [comment]
Will 4.2 (or later) the one released in WP 5.0? Asking because this is a regression and I'd expect it to be fixed for release.

**Comment by talldan** [comment]
😄 Sure thing, I'll take a look.

**Comment by talldan** [comment]
Hi @afercia & @abrightclearweb

I did some testing and agree, this is far from easy for a keyboard user. I have a feeling the link popover visibility is tied to that of the toolbar. I had to use shift & left-arrow to select some text. Then I could tab to the popover, but only after going through all the sidebar elements.

Just to make sure I work on the right thing, the desired behaviour would be
1. The link popover is visible whenever the caret is positioned inside the link text.
2. Pressing tab once switches focus to the popover (first the link, second the edit button, and third the settings button).

Does that sound correct?

**Comment by afercia** [comment]
@talldan thanks for looking at this.
I'd say the Tab behavior shouldn't change, that wouldn't be expected at this point given the popover is placed far in the DOM.

> I had to use shift & left-arrow to select some link text. Then I could tab to the popover, but only after going through all the sidebar elements.

I guess you've tested with just one paragraph in the post. When there are other blocks, tabbing brings you to the following block. At that point, the previous block with the link is no more "selected" and the link toolbar disappears. 🙂So we can't rely on tabbing.

I see something has changed in current master, specifically this point in the original report:

> pressing `Cmd+K` makes the link toolbar appear but it's in "insert mode" and doesn't allow me to modify the existing link

Instead, now `Cmd+K` on an existing link opens the link toolbar in "edit mode". This solves great part of the problem, as the link is now editable.

Seems to me the only problem left compared to the previous behavior with the Classic Editor is that there's the need to select _all the link text_ to make `Cmd+K` work and open the link toolbar. Instead, on Classic, having the caret positioned within the link is enough to make `Cmd+K` work.

Aside: since we're on it, can we change `aria-label="URL"` to `aria-label="Link URL"` ?

**Comment by talldan** [comment]
> I guess you've tested with just one paragraph in the post. When there are other blocks, tabbing brings you to the following block. At that point, the previous block with the link is no more "selected" and the link toolbar disappears. 🙂So we can't rely on tabbing.

Good point 🤦‍♂️ 

> Instead, on Classic, having the caret positioned within the link is enough to make Cmd+K work.

As suspected, this does seem to be tied to the visibility of the toolbar. The link popover is technically rendered within the toolbar (though within a portal, so it appears elsewhere in the DOM). That means it is only displayed when the toolbar is displayed. Here I hacked the toolbar so that it's always displayed and hey presto, the link popover is also displayed whenever the caret is within the link:

![oct-19-2018 16-20-16](https://user-images.githubusercontent.com/677833/47206382-15082300-d3bb-11e8-850a-bcedf020ffa4.gif)

**Comment by talldan** [comment]
So I can think of two potential solutions
1) Make sure the link popover isn't rendered as a child of the toolbar. This means the popover will be rendered, but not the toolbar. That looks like this:
![oct-19-2018 16-37-23](https://user-images.githubusercontent.com/677833/47207249-56013700-d3bd-11e8-9a52-82055909b69c.gif)

2) Display the toolbar whenever the caret is within some formatting, thus causing the popover to also be rendered.

The second one would probably require a little bit more work. I think this has turned into a design question, so I'll add the Needs Design label.

**Comment by talldan** [comment]
If there's no answer next week, I'll go ahead an put together a PR of option 1.

> Aside: since we're on it, can we change aria-label="URL" to aria-label="Link URL" ?

I'll make sure to include that 👍

**Comment by talldan** [comment]
I've put together a PR for the second option #10983, would be great to have some testing. I also have a branch locally for the first option.

I'm going to sort out a couple of things on separate PRs to make it easier to review and merge:
- Changing the aria-label.
- Making the sure the url popover stays centered (so that it doesn't follow the caret sideways.)
